<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Владислав Станин">
<meta name="dcterms.date" content="2024-11-22">

<title>Gene Prediction – POLYTECH BIOINF</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="../styles.css">
</head>

<body class="nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">POLYTECH BIOINF</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../index.html"> 
<span class="menu-text">Home</span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#данные" id="toc-данные" class="nav-link active" data-scroll-target="#данные">Данные</a></li>
  <li><a href="#перед-идентификацией-генов-поговорим-о-чём-помельче" id="toc-перед-идентификацией-генов-поговорим-о-чём-помельче" class="nav-link" data-scroll-target="#перед-идентификацией-генов-поговорим-о-чём-помельче">Перед идентификацией генов поговорим о чём помельче…</a></li>
  <li><a href="#repeats-and-transposons" id="toc-repeats-and-transposons" class="nav-link" data-scroll-target="#repeats-and-transposons">Repeats and transposons</a></li>
  <li><a href="#repeats-and-transposons-1" id="toc-repeats-and-transposons-1" class="nav-link" data-scroll-target="#repeats-and-transposons-1">Repeats and transposons</a></li>
  <li><a href="#repeatmasker" id="toc-repeatmasker" class="nav-link" data-scroll-target="#repeatmasker">RepeatMasker</a></li>
  <li><a href="#trna" id="toc-trna" class="nav-link" data-scroll-target="#trna">tRNA</a></li>
  <li><a href="#trna-1" id="toc-trna-1" class="nav-link" data-scroll-target="#trna-1">tRNA</a></li>
  <li><a href="#rrna" id="toc-rrna" class="nav-link" data-scroll-target="#rrna">rRNA</a></li>
  <li><a href="#rna-from-rfam-via-hmmer3" id="toc-rna-from-rfam-via-hmmer3" class="nav-link" data-scroll-target="#rna-from-rfam-via-hmmer3">RNA from Rfam via HMMER3</a></li>
  <li><a href="#подбираемся-к-белок-кодирующим-генам" id="toc-подбираемся-к-белок-кодирующим-генам" class="nav-link" data-scroll-target="#подбираемся-к-белок-кодирующим-генам">Подбираемся к белок-кодирующим генам</a></li>
  <li><a href="#open-reading-frame" id="toc-open-reading-frame" class="nav-link" data-scroll-target="#open-reading-frame">Open reading frame</a></li>
  <li><a href="#open-reading-frame-1" id="toc-open-reading-frame-1" class="nav-link" data-scroll-target="#open-reading-frame-1">Open reading frame</a>
  <ul class="collapse">
  <li><a href="#ncbi-online-tool" id="toc-ncbi-online-tool" class="nav-link" data-scroll-target="#ncbi-online-tool">NCBI online tool</a></li>
  <li><a href="#utilities-from-emboss" id="toc-utilities-from-emboss" class="nav-link" data-scroll-target="#utilities-from-emboss">Utilities from EMBOSS:</a></li>
  </ul></li>
  <li><a href="#coding-sequence-cds" id="toc-coding-sequence-cds" class="nav-link" data-scroll-target="#coding-sequence-cds">Coding sequence (CDS)</a></li>
  <li><a href="#coding-sequence-cds-1" id="toc-coding-sequence-cds-1" class="nav-link" data-scroll-target="#coding-sequence-cds-1">Coding sequence (CDS)</a></li>
  <li><a href="#coding-sequence-cds-2" id="toc-coding-sequence-cds-2" class="nav-link" data-scroll-target="#coding-sequence-cds-2">Coding sequence (CDS)</a></li>
  <li><a href="#genemark" id="toc-genemark" class="nav-link" data-scroll-target="#genemark">GeneMark</a></li>
  <li><a href="#augustus" id="toc-augustus" class="nav-link" data-scroll-target="#augustus">Augustus</a></li>
  <li><a href="#augustus-1" id="toc-augustus-1" class="nav-link" data-scroll-target="#augustus-1">Augustus</a></li>
  <li><a href="#augustus-2" id="toc-augustus-2" class="nav-link" data-scroll-target="#augustus-2">Augustus</a></li>
  <li><a href="#busco" id="toc-busco" class="nav-link" data-scroll-target="#busco">BUSCO</a></li>
  <li><a href="#busco-1" id="toc-busco-1" class="nav-link" data-scroll-target="#busco-1">BUSCO</a></li>
  <li><a href="#busco-2" id="toc-busco-2" class="nav-link" data-scroll-target="#busco-2">BUSCO</a></li>
  <li><a href="#busco-3" id="toc-busco-3" class="nav-link" data-scroll-target="#busco-3">BUSCO</a></li>
  <li><a href="#busco-output" id="toc-busco-output" class="nav-link" data-scroll-target="#busco-output">BUSCO output</a></li>
  <li><a href="#busco-output-1" id="toc-busco-output-1" class="nav-link" data-scroll-target="#busco-output-1">BUSCO output</a></li>
  <li><a href="#busco-output-2" id="toc-busco-output-2" class="nav-link" data-scroll-target="#busco-output-2">BUSCO output</a></li>
  <li><a href="#deep-learning-approaches" id="toc-deep-learning-approaches" class="nav-link" data-scroll-target="#deep-learning-approaches">Deep Learning approaches</a></li>
  <li><a href="#deep-learning-approaches-1" id="toc-deep-learning-approaches-1" class="nav-link" data-scroll-target="#deep-learning-approaches-1">Deep Learning approaches</a></li>
  <li><a href="#deep-learning-approaches-2" id="toc-deep-learning-approaches-2" class="nav-link" data-scroll-target="#deep-learning-approaches-2">Deep Learning approaches</a></li>
  <li><a href="#homework" id="toc-homework" class="nav-link" data-scroll-target="#homework">Homework</a></li>
  </ul>
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="gene_prediction.html"><i class="bi bi-file-slides"></i>RevealJS</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Gene Prediction</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Владислав Станин </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">November 22, 2024</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="данные" class="level2">
<h2 class="anchored" data-anchor-id="данные">Данные</h2>
<p>Сегодня будем всячески аннотировать хромосому непатогенного гриба <strong>Fusarium oxysporum Fo47.</strong></p>
<div class="fragment fade-in">
<p>Файл <code>chr_fo47.fasta</code></p>
</div>
</section>
<section id="перед-идентификацией-генов-поговорим-о-чём-помельче" class="level2" data-transition="slide-out">
<h2 data-transition="slide-out" class="anchored" data-anchor-id="перед-идентификацией-генов-поговорим-о-чём-помельче">Перед идентификацией генов поговорим о чём помельче…</h2>
</section>
<section id="repeats-and-transposons" class="level2" data-auto-animate="true" style="text-align: center">
<h2 data-auto-animate="true" style="text-align: center" class="anchored" data-anchor-id="repeats-and-transposons">Repeats and transposons</h2>
</section>
<section id="repeats-and-transposons-1" class="level2" data-auto-animate="true" style="text-align: center">
<h2 data-auto-animate="true" style="text-align: center" class="anchored" data-anchor-id="repeats-and-transposons-1">Repeats and transposons</h2>
<p>Повторы могут мешать при аннотации генома, пораждая <strong>ложные гены и открытые рамки</strong>.</p>
<div class="fragment fade-in">
<p>Поэтому мы их замаскируем - поменяем на <code>NNNNN</code></p>
</div>
</section>
<section id="repeatmasker" class="level2">
<h2 class="anchored" data-anchor-id="repeatmasker">RepeatMasker</h2>
<p>На основе известных последовательностей</p>
<p><em>Самый простой способ:</em></p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">RepeatMasker</span> <span class="op">&lt;</span>fasta-file<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>или (смотря откуда скачивали)</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">repeatmasker</span> <span class="op">&lt;</span>fasta-file<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><code>--xsmall</code> для <em>softmasking</em></p>
<div class="fragment fade-in">
<ul>
<li><strong>Удалит и малые РНК</strong> - осторожно!</li>
<li>Можно использовать базу Dfam цкликом (8.9Gb)</li>
<li>Предсказывать транспозоны <em>ab initio</em>: <strong>Repeatmodeler</strong></li>
</ul>
</div>
</section>
<section id="trna" class="level2" data-auto-animate="true">
<h2 data-auto-animate="true" class="anchored" data-anchor-id="trna">tRNA</h2>
<p><strong>tRNAScan-SE</strong></p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">trnascan-se</span> <span class="at">-E</span> <span class="op">&lt;</span>input fasta<span class="op">&gt;</span> -o <span class="op">&lt;</span>output<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><strong>или</strong> (смотря откуда вы скачивали)</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="ex">tRNAscan-SE</span> <span class="at">-E</span> <span class="op">&lt;</span>input fasta<span class="op">&gt;</span> -o <span class="op">&lt;</span>output<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><code>-E</code> - обозначает поиск эукариотических тРНК</p>
</section>
<section id="trna-1" class="level2" data-auto-animate="true">
<h2 data-auto-animate="true" class="anchored" data-anchor-id="trna-1">tRNA</h2>
<p><strong>tRNAScan-SE</strong></p>
<pre><code>  -E         : search for eukaryotic tRNAs (default)
  -B         : search for bacterial tRNAs
  -A         : search for archaeal tRNAs
  -M &lt;model&gt; : search for mitochondrial tRNAs
                 options: mammal, vert
  -O         : search for other organellar tRNAs
  -G         : use general tRNA model (cytoslic tRNAs from all 3 domains included)</code></pre>
</section>
<section id="rrna" class="level2">
<h2 class="anchored" data-anchor-id="rrna">rRNA</h2>
<p><strong>barrnap</strong></p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">barrnap</span> <span class="at">--kingdom</span> euk <span class="op">&lt;</span>fasta file<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><code>--kingdom euk</code> - чтобы искать <strong>28S, 18S, 5.8S, 5S</strong></p>
<div class="fragment fade-in">
<p><code>--kingdom [X]</code> Kingdom: <em>euk, bac, arc, mito</em> <strong>(default ‘bac’)</strong></p>
</div>
</section>
<section id="rna-from-rfam-via-hmmer3" class="level2">
<h2 class="anchored" data-anchor-id="rna-from-rfam-via-hmmer3">RNA from Rfam via HMMER3</h2>
<div class="sourceCode" id="cb7"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb7-1"><a href="#cb7-1"></a><span class="fu">wget</span> https://ftp.ebi.ac.uk/pub/databases/Rfam/CURRENT/Rfam.seed.gz</span>
<span id="cb7-2"><a href="#cb7-2"></a><span class="fu">gunzip</span> <span class="at">-k</span> Rfam.seed.gz</span>
<span id="cb7-3"><a href="#cb7-3"></a></span>
<span id="cb7-4"><a href="#cb7-4"></a><span class="ex">hmmbuild</span> Rfam.seed.hmm Rfam.seed</span>
<span id="cb7-5"><a href="#cb7-5"></a><span class="ex">hmmpress</span> Rfam.seed.hmm</span>
<span id="cb7-6"><a href="#cb7-6"></a><span class="ex">nhmmscan</span> <span class="at">--tblout</span> rnas_result.txt Rfam.seed.hmm <span class="op">&lt;</span>fasta file<span class="op">&gt;</span> </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Кстати, <code>Rfam.seed</code> - не FASTA!</p>
</section>
<section id="подбираемся-к-белок-кодирующим-генам" class="level2" style="text-align: center" data-transition="slide-out">
<h2 style="text-align: center" data-transition="slide-out" class="anchored" data-anchor-id="подбираемся-к-белок-кодирующим-генам">Подбираемся к белок-кодирующим генам</h2>
</section>
<section id="open-reading-frame" class="level2" data-auto-animate="true" style="text-align: center">
<h2 data-auto-animate="true" style="text-align: center" class="anchored" data-anchor-id="open-reading-frame">Open reading frame</h2>
</section>
<section id="open-reading-frame-1" class="level2" data-auto-animate="true" style="text-align: center">
<h2 data-auto-animate="true" style="text-align: center" class="anchored" data-anchor-id="open-reading-frame-1">Open reading frame</h2>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="_files/gene_prediction/orf.jpeg" class="img-fluid quarto-figure quarto-figure-center figure-img" width="750"></p>
</figure>
</div>
<p>Самое простое представление гена (но неверное):<br>
<strong>Последовательность триплетов от старт- до стоп-кодона</strong></p>
<hr>
<section id="ncbi-online-tool" class="level3">
<h3 class="anchored" data-anchor-id="ncbi-online-tool">NCBI online tool</h3>
<p><a href="https://www.ncbi.nlm.nih.gov/orffinder/"><strong>ORFfinder</strong></a></p>
</section>
<section id="utilities-from-emboss" class="level3">
<h3 class="anchored" data-anchor-id="utilities-from-emboss">Utilities from EMBOSS:</h3>
<p><strong>getorf</strong></p>
<div class="sourceCode" id="cb8"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="ex">getorf</span> <span class="at">-sequence</span> <span class="op">&lt;</span>inputfile<span class="op">&gt;</span> -outseq <span class="op">&lt;</span>outputfile<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="fragment fade-in">
<p>Longest ORF (3 штуки):</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="ex">seqkit</span> fx2tab <span class="at">--length</span> <span class="op">&lt;</span>orf fasta file<span class="op">&gt;</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>   <span class="fu">sort</span> <span class="at">-t</span> <span class="st">$'</span><span class="dt">\t</span><span class="st">'</span> <span class="at">-k4,4nr</span> <span class="kw">|</span> <span class="dt">\</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>   <span class="fu">head</span> <span class="at">-n</span> 3 <span class="kw">|</span> <span class="ex">seqkit</span> tab2fx <span class="op">&gt;</span> longest_orf.fasta</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<!-- #### **transeq** -->
<!-- ``` bash -->
<!-- transeq -sequence <input ORF file> \ -->
<!--    -outseq <outputfile> -->
<!-- ``` -->
<hr>
<ul>
<li><p><u><strong>ORF - не ген, а просто открытая рамка, которая может найтись даже случано</strong></u></p></li>
<li><p><strong>Не учитывает интроны</strong> в ДНК (плох для эукариот)</p></li>
<li><p>Важно выбрать специфический <strong>генетический код</strong></p></li>
<li><p>Можно использовать для мРНК (нет интронов, но проблемы с UTR)</p></li>
</ul>
</section>
</section>
<section id="coding-sequence-cds" class="level2" data-auto-animate="true" style="text-align: center">
<h2 data-auto-animate="true" style="text-align: center" class="anchored" data-anchor-id="coding-sequence-cds">Coding sequence (CDS)</h2>
</section>
<section id="coding-sequence-cds-1" class="level2" data-auto-animate="true" style="text-align: center">
<h2 data-auto-animate="true" style="text-align: center" class="anchored" data-anchor-id="coding-sequence-cds-1">Coding sequence (CDS)</h2>
<p>Отражает кодирующие области гена (экзоны без UTR) и альтреанативный сплайсинг.</p>
</section>
<section id="coding-sequence-cds-2" class="level2" data-auto-animate="true" style="text-align: center">
<h2 data-auto-animate="true" style="text-align: center" class="anchored" data-anchor-id="coding-sequence-cds-2">Coding sequence (CDS)</h2>
<p>Отражает кодирующие области гена (экзоны без UTR) и альтреанативный сплайсинг.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="_files/gene_prediction/gene_struct.jpg" class="img-fluid quarto-figure quarto-figure-center figure-img" width="1400"></p>
</figure>
</div>
<p>(Синие области - <strong>CDS</strong>)</p>
</section>
<section id="genemark" class="level2">
<h2 class="anchored" data-anchor-id="genemark">GeneMark</h2>
<p>Семейство программ для разметки генов.</p>
<div class="columns">
<div class="column" style="width:38%;">
<ol type="1">
<li><p><strong>GeneMark.hmm</strong></p>
<ul>
<li><a href="https://genemark.bme.gatech.edu/GeneMark/gmhmmp.cgi">прокариоты</a></li>
<li><a href="https://genemark.bme.gatech.edu/GeneMark/gmhmme.cgi">эукариоты</a></li>
<li><a href="https://genemark.bme.gatech.edu/GeneMark/meta_gmhmmp.cgi"><em>метагеномы</em></a></li>
</ul></li>
</ol>
</div><div class="column" style="width:62%;">
<ol start="2" type="1">
<li><a href="https://genemark.bme.gatech.edu/GeneMark/genemarks2.cgi"><strong>GeneMark-S2</strong></a> - <em>self-training</em> и только прокариоты, археи</li>
</ol>
<p>Все версии <a href="https://genemark.bme.gatech.edu/GeneMark/">ТУТ</a>, включая новейшие standalone версии</p>
</div>
</div>
<p><strong>Важно выбрать модель организма наиболее близкого к вашему</strong></p>
</section>
<section id="augustus" class="level2" data-auto-animate="true">
<h2 data-auto-animate="true" class="anchored" data-anchor-id="augustus">Augustus</h2>
<div class="sourceCode" id="cb10"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb10-1"><a href="#cb10-1"></a><span class="ex">augustus</span> <span class="dt">\</span></span>
<span id="cb10-2"><a href="#cb10-2"></a>   <span class="at">--protein</span><span class="op">=</span>on <span class="dt">\ </span># предсказать кодируемые белки</span>
<span id="cb10-3"><a href="#cb10-3"></a>   <span class="ex">--hintsfile=</span><span class="op">&lt;</span>gff-file<span class="op">&gt;</span> <span class="dt">\ </span># <span class="st">"подсказки"</span>-гены</span>
<span id="cb10-4"><a href="#cb10-4"></a>   <span class="ex">--genemodel=complete</span> <span class="dt">\ </span># только полные гены</span>
<span id="cb10-5"><a href="#cb10-5"></a>   <span class="ex">--species=human</span> <span class="dt">\ </span># надо выбрать species!</span>
<span id="cb10-6"><a href="#cb10-6"></a>   <span class="ex">fasta.file</span> <span class="op">&gt;</span> output.gff</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Список параметров и доступных species: <a href="https://github.com/Gaius-Augustus/Augustus/blob/master/docs/RUNNING-AUGUSTUS.md#basic-usage">документация</a></p>
<p>Готовые подсказки - это gff-файл, где часть структур уже размечено (optional)</p>
<p><code>--genemodel</code> по умолчанию <strong>partial</strong> - находит даже частичные гены</p>
</section>
<section id="augustus-1" class="level2" data-auto-animate="true">
<h2 data-auto-animate="true" class="anchored" data-anchor-id="augustus-1">Augustus</h2>
<p>Если вашего species нету, то можно <strong>обучить</strong> Augustus, используя известные гены родственного организма. Например, из ENSEMBL или даже с <a href="http://bioinf.uni-greifswald.de/augustus/datasets/">сайта Augustus</a></p>
<p><a href="https://bioinf.uni-greifswald.de/augustus/binaries/tutorial/training.html"><strong>Туториал</strong></a></p>
</section>
<section id="augustus-2" class="level2" data-auto-animate="true">
<h2 data-auto-animate="true" class="anchored" data-anchor-id="augustus-2">Augustus</h2>
<p>Содержит в себе много полезных скриптов, например</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="ex">getAnnoFasta.pl</span> <span class="op">&lt;</span>augustus.gff<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>создает FASTA с предсказанными Augustus белками</p>
</section>
<section id="busco" class="level2">
<h2 class="anchored" data-anchor-id="busco">BUSCO</h2>
<p>Benchmarking Universal <strong>Single-Copy Orthologs</strong> - база данных с говорящим названием. Чаще используется для оценки “полноты” <strong>полного генома</strong></p>
<div class="sourceCode" id="cb12"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1"></a><span class="ex">busco</span> <span class="at">-i</span> <span class="op">&lt;</span>fasta file<span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb12-2"><a href="#cb12-2"></a>   <span class="at">-m</span> genome <span class="dt">\</span></span>
<span id="cb12-3"><a href="#cb12-3"></a>   <span class="at">-l</span> <span class="op">&lt;</span>lineage<span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb12-4"><a href="#cb12-4"></a>   <span class="at">--cpu</span> 8 <span class="dt">\</span></span>
<span id="cb12-5"><a href="#cb12-5"></a>   <span class="at">-o</span> <span class="op">&lt;</span>output directory<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p><code>-m</code> - можно указать <strong>protein</strong> - аннотация будет по протеому - тогда для <code>-i</code> надо указать <em>fasta</em> с белками</p>
<p><strong>Работает только с VPN или локально скачанными базами lineage</strong></p>
</section>
<section id="busco-1" class="level2" data-auto-animate="true">
<h2 data-auto-animate="true" class="anchored" data-anchor-id="busco-1">BUSCO</h2>
<p>При первых запусках новые lineage будут скачиваться локально. Но этого не проивойдет без VPN… Для запуска оффлайн нужно скачать нужные базы и положить их в директорию опеределнным образом:</p>
<pre><code>busco_downloads/
├── file_versions.tsv
└── lineages
    ├── ascomycota_odb10
    ├── fungi_odb10
    └── hypocreales_odb10</code></pre>
<p>базы odb надо скачать… с VPN… <a href="https://busco-data.ezlab.org/v5/data/"><strong>отсюда</strong></a> или попросить у товарища</p>
</section>
<section id="busco-2" class="level2" data-auto-animate="true">
<h2 data-auto-animate="true" class="anchored" data-anchor-id="busco-2">BUSCO</h2>
<p>Запускаем оффлайн</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1"></a><span class="ex">busco</span> <span class="at">-i</span> <span class="op">&lt;</span>fasta file<span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb14-2"><a href="#cb14-2"></a>   <span class="at">-m</span> genome <span class="dt">\</span></span>
<span id="cb14-3"><a href="#cb14-3"></a>   <span class="at">-l</span> <span class="op">&lt;</span>lineage<span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb14-4"><a href="#cb14-4"></a>   <span class="at">--offline</span> <span class="dt">\</span></span>
<span id="cb14-5"><a href="#cb14-5"></a>   <span class="at">--download_path</span> <span class="op">&lt;</span>path/to/directory<span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb14-6"><a href="#cb14-6"></a>   <span class="at">--cpu</span> 8 <span class="dt">\</span></span>
<span id="cb14-7"><a href="#cb14-7"></a>   <span class="at">-o</span> <span class="op">&lt;</span>output directory<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Скачанный lineage лучше указывать не как <strong>fusarium</strong>, а как <strong>fusarium_odb10</strong> - то есть навзание папки, так как версии lineage могут отличаться.</p>
</section>
<section id="busco-3" class="level2" data-transition="fade-out">
<h2 data-transition="fade-out" class="anchored" data-anchor-id="busco-3">BUSCO</h2>
<p>Benchmarking Universal <strong>Single-Copy Orthologs</strong></p>
<div class="sourceCode" id="cb15"><pre class="sourceCode numberSource bash number-lines code-with-copy"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1"></a><span class="ex">busco</span> <span class="at">-i</span> <span class="op">&lt;</span>fasta file<span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb15-2"><a href="#cb15-2"></a>   <span class="at">-m</span> <span class="op">&lt;</span>genome/protein<span class="op">&gt;</span> <span class="dt">\</span></span>
<span id="cb15-3"><a href="#cb15-3"></a>   <span class="at">--auto-lineage</span> <span class="dt">\ </span># если не уверены какой lineage</span>
<span id="cb15-4"><a href="#cb15-4"></a>   <span class="ex">--cpu</span> 8 <span class="dt">\</span></span>
<span id="cb15-5"><a href="#cb15-5"></a>   <span class="at">-o</span> <span class="op">&lt;</span>output directory<span class="op">&gt;</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Но это не панацея - <strong>он ошибается</strong></p>
<div class="fragment fade-in">
<ul>
<li><code>--auto-lineage-euk</code> - среди эукариот</li>
<li><code>--auto-lineage-prok</code>- среди прокариот</li>
</ul>
</div>
</section>
<section id="busco-output" class="level2" data-auto-animate="true">
<h2 data-auto-animate="true" class="anchored" data-anchor-id="busco-output">BUSCO output</h2>
<p>Пусть команда была:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="ex">busco</span> <span class="at">-i</span> chr_fo47.fasta <span class="at">-l</span> fungi_odb10 <span class="at">--offline</span> <span class="dt">\</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>   <span class="at">--download_path</span> busco_downloads <span class="dt">\</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>   <span class="at">-m</span> genome <span class="at">-o</span> busco_output_fungi</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<p>Папка <code>busco_output_fungi/run_fungi_odb10/</code>:</p>
<pre><code>busco_output_fungi/run_fungi_odb10/
├── busco_sequences
├── full_table.tsv
├── hmmer_output
├── miniprot_output
├── missing_busco_list.tsv
├── short_summary.json
└── short_summary.txt</code></pre>
</section>
<section id="busco-output-1" class="level2" data-auto-animate="true">
<h2 data-auto-animate="true" class="anchored" data-anchor-id="busco-output-1">BUSCO output</h2>
<p>Папка <code>busco_output_fungi/run_fungi_odb10/</code>:</p>
<pre><code>busco_output_fungi/run_fungi_odb10/
├── busco_sequences
├── full_table.tsv
├── hmmer_output
├── miniprot_output
├── missing_busco_list.tsv
├── short_summary.json
└── short_summary.txt</code></pre>
<p>Интереснее всего в <em>full_table.tsv</em> - тут аннотация генов и <em>short_summary.txt</em> - тут общая статистика и <strong>completeness</strong></p>
</section>
<section id="busco-output-2" class="level2" data-auto-animate="true">
<h2 data-auto-animate="true" class="anchored" data-anchor-id="busco-output-2">BUSCO output</h2>
<p>Интереснее всего в <em>full_table.tsv</em> - тут аннотация генов и <em>short_summary.txt</em> - тут общая статистика и <strong>completeness</strong></p>
<p>Пример одной и строчек <em>short_summary.txt</em>:</p>
<pre class="text"><code>C:11.2%[S:11.2%,D:0.0%],F:0.4%,M:88.4%,n:758,E:11.8%</code></pre>
<ul>
<li>C:11.2% - <strong>completeness</strong></li>
<li>[S:11.2%,D:0.0%] - однокопийных 11.2% и 0% дупликатов</li>
<li>F:0.4% - Fragmented</li>
<li>M:88.4% - Missed</li>
</ul>
</section>
<section id="deep-learning-approaches" class="level2" data-auto-animate="true" style="text-align: center">
<h2 data-auto-animate="true" style="text-align: center" class="anchored" data-anchor-id="deep-learning-approaches">Deep Learning approaches</h2>
</section>
<section id="deep-learning-approaches-1" class="level2" data-auto-animate="true" style="text-align: center">
<h2 data-auto-animate="true" style="text-align: center" class="anchored" data-anchor-id="deep-learning-approaches-1">Deep Learning approaches</h2>
<p>Используют нейронные сети (обычно свёрточные) и машинное обучение для предсказания</p>
<div class="columns">
<div class="column" style="width:50%;">
<p><strong>Tiberius</strong></p>
</div><div class="column" style="width:50%;">
<p><strong>Helixer</strong></p>
</div>
</div>
</section>
<section id="deep-learning-approaches-2" class="level2" data-auto-animate="true" style="text-align: center">
<h2 data-auto-animate="true" style="text-align: center" class="anchored" data-anchor-id="deep-learning-approaches-2">Deep Learning approaches</h2>
<div class="columns">
<div class="column" style="width:50%;">
<p><strong>Tiberius</strong></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="_files/gene_prediction/tiberius.png" class="img-fluid quarto-figure quarto-figure-center figure-img"></p>
</figure>
</div>
</div><div class="column" style="width:50%;">
<p><strong>Helixer</strong></p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="_files/gene_prediction/helixer.png" class="img-fluid quarto-figure quarto-figure-center figure-img" width="1500"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="homework" class="level2">
<h2 class="anchored" data-anchor-id="homework">Homework</h2>
<p>Аннотировать хромосомы пивных дрожжей</p>
<div class="columns">
<div class="column" style="width:50%;">
<p><img src="_files/gene_prediction/kvas.jpg" class="img-fluid" width="570"></p>
</div><div class="column" style="width:50%;">
<p><img src="_files/gene_prediction/yeasts.jpg" class="img-fluid" width="612"></p>
</div>
</div>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>